teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -0,0 +1,205 @@
+stops all running tasks and deletes associated Pods. That prevents final tasks, specified in a `Pipeline` under `finally` section,
+from being run.
+There should be a way to terminate a `PipelineRun` gracefully and wait for cleanup actions triggered by final tasks.
+
+## Motivation
+
+It is common that tasks when run trigger execution of external activities or request resources in foreign systems.
+Final tasks are perfectly suited for any cleanup operations that have to be performed when execution is completed.
+That is required in case of success, failure, as well as pipeline run cancellation.
+ 
+Currently, there is no way to terminate a `PipelineRun` gracefully.
+The existing cancellation capability is problematic in the real use cases.
+
+This is especially important for users of [Kubeflow Pipelines with Tekton backend](https://github.com/kubeflow/kfp-tekton/).
+Kubeflow Pipelines supports [exit handler](https://www.kubeflow.org/docs/components/pipelines/overview/pipelines-overview/)
+which guarantees that selected operations are triggered whenever pipeline run is completed.
+Those actions are executed in case of a pipeline run cancellation as well. 
+Given the fact that Kubeflow Pipelines' `ExitHandler` is implemented in Tekton using final tasks,
+there is a significant inconsistency in the describe behaviour.
+
+Lastly, final tasks should be triggered on a pipeline run timeout, which is a standard error scenario.
+Running final tasks infinitely should be prevented with the additional configuration of a finalization timeout.
+There is a separate proposal: [TEP-0047](https://github.com/tektoncd/community/pull/326) that covers this part.
+
+Related issues:
+- https://github.com/kubeflow/kfp-tekton/issues/506
+- https://github.com/tektoncd/community/pull/326
+
+
+### Goals
+
+- Add support for graceful a pipeline run termination / stop, that would wait for final tasks to be completed before
+  termination.
+
+### Non-Goals
+
+- There is not intention to change the existing pipeline run cancellation behaviour, 
+  but rather provide an alternative one that would support graceful run termination.
+
+### Use Cases (optional)
+
+- As a Kubeflow user, I run a training pipeline that spawns a [`TFJob`](https://www.kubeflow.org/docs/components/training/tftraining/).
+  Once the pipeline execution is stopped the training job should be terminated to limit resource usage.
+- As a CodeEngine user, I run a pipeline that [submits a batch job](https://cloud.ibm.com/docs/codeengine?topic=codeengine-cli#cli-jobrun).
+  Once the pipeline execution is stopped the batch job should be terminated to limit the service costs.
+- As a kubernetes user, I run a pipeline that provisions a new cluster.
+  The pipeline is executed and stopped. The k8s cluster resource should be freed up.
+
+## Requirements
+
+- Users should be able to gracefully stop a pipeline run and cleanup external resources. 
+
+## Proposal
```

Comment:
@rafalbigaj proposal is missing some details on what should be the status of terminating `taskRuns`. I would recommend dropping the proposal for now.

We can merge the problem statement first without the proposal. We have been following this approach to move faster with the review process.

Reply:
Done.

Reply:
When it's time to commit the proposal, we also need to think how should this new state impact the `pipelineRun` if it's already running `finally` tasks. I am assuming, there should not be any impact on any `finally` task i.e. `pipelineRun` would ignore this new state and continue running `finally` until finish.